File: LP/practical_1_bfs_dfs/bfs_dfs.py
from collections import deque

class Graph:
    def __init__(self, vertices):
        self.vertices = vertices
        self.adj_list = [[] for _ in range(vertices)]

    def add_edge(self, u, v):
        self.adj_list[u].append(v)
        self.adj_list[v].append(u)

    # Recursive DFS
    def dfs_recursive(self, v, visited):
        visited[v] = True
        print(v, end=' ')

        for neighbor in self.adj_list[v]:
            if not visited[neighbor]:
                self.dfs_recursive(neighbor, visited)

    def dfs(self, start):
        visited = [False] * self.vertices
        print("DFS Traversal (Recursive): ", end='')
        self.dfs_recursive(start, visited)
        print()

    # BFS using queue
    def bfs(self, start):
        visited = [False] * self.vertices
        queue = deque()

        visited[start] = True
        queue.append(start)

        print("BFS Traversal: ", end='')

        while queue:
            v = queue.popleft()
            print(v, end=' ')

            for neighbor in self.adj_list[v]:
                if not visited[neighbor]:
                    visited[neighbor] = True
                    queue.append(neighbor)
        print()

# Main program
def main():
    V = int(input("Enter number of vertices: "))
    E = int(input("Enter number of edges: "))

    g = Graph(V)

    print("Enter edges (u v):")
    for _ in range(E):
        u, v = map(int, input().split())
        g.add_edge(u, v)

    start = int(input("Enter starting vertex for traversal: "))
    g.dfs(start)
    g.bfs(start)

if __name__ == "__main__":
    main()

File: LP/practical_1_bfs_dfs/bfs_dfs.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

class Graph {
private:
    int vertices;
    vector<vector<int>> adjList;

public:
    Graph(int V) {
        vertices = V;
        adjList.resize(V);
    }

    void addEdge(int u, int v) {
        adjList[u].push_back(v);
        adjList[v].push_back(u);
    }

    // Recursive DFS function
    void DFSRecursive(int v, vector<bool>& visited) {
        visited[v] = true;
        cout << v << " ";

        for (int neighbor : adjList[v]) {
            if (!visited[neighbor]) {
                DFSRecursive(neighbor, visited);
            }
        }
    }

    // DFS traversal (recursive)
    void DFS(int start) {
        vector<bool> visited(vertices, false);
        cout << "DFS Traversal (Recursive): ";
        DFSRecursive(start, visited);
        cout << endl;
    }

    // BFS traversal (using queue)
    void BFS(int start) {
        vector<bool> visited(vertices, false);
        queue<int> q;

        visited[start] = true;
        q.push(start);

        cout << "BFS Traversal: ";

        while (!q.empty()) {
            int v = q.front();
            q.pop();
            cout << v << " ";

            for (int neighbor : adjList[v]) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    q.push(neighbor);
                }
            }
        }

        cout << endl;
    }
};

int main() {
    int V, E;
    cout << "Enter number of vertices: ";
    cin >> V;
    cout << "Enter number of edges: ";
    cin >> E;

    Graph g(V);

    cout << "Enter edges (u v):" << endl;
    for (int i = 0; i < E; i++) {
        int u, v;
        cin >> u >> v;
        g.addEdge(u, v);
    }

    int startVertex;
    cout << "Enter starting vertex for traversal: ";
    cin >> startVertex;

    g.DFS(startVertex); // Recursive DFS
    g.BFS(startVertex); // BFS

    return 0;
}

